pkg/microservice/systemconfig/core/codehost/service/codehost.go

```diff
@@ -69,3 +78,114 @@ func UpdateCodeHostByToken(host *models.CodeHost, _ *zap.SugaredLogger) (*models
+		return "", err
+	}
+	redirectParsedURL, err := url.Parse(redirectURI)
+	if err != nil {
+		logger.Errorf("Parse redirectURI:%s err:%s", redirectURI, err)
+		return "", err
+	}
+	callbackURL := url.URL{
+		Scheme: redirectParsedURL.Scheme,
+		Host:   redirectParsedURL.Host,
+		Path:   callback,
+	}
+	oauth, err := newOAuth(codeHost.Type, callbackURL.String(), codeHost.ApplicationId, codeHost.ClientSecret, codeHost.Address)
+	if err != nil {
+		logger.Errorf("NewOAuth:%s err:%s", codeHost.Type, err)
+		return "", err
+	}
+	stateStruct := state{
+		CodeHostID:  codeHost.ID,
+		RedirectURL: redirectURI,
+	}
+	bs, err := json.Marshal(stateStruct)
+	if err != nil {
+		logger.Errorf("Marshal err:%s", err)
+		return "", err
+	}
+	return oauth.LoginURL(base64.URLEncoding.EncodeToString(bs)), nil
+}
+
+func HandleCallback(stateStr string, r *http.Request, logger *zap.SugaredLogger) (string, error) {
+	// TODO：validate the code
+	// https://www.jianshu.com/p/c7c8f51713b6
+	decryptedState, err := base64.URLEncoding.DecodeString(stateStr)
+	if err != nil {
+		logger.Errorf("DecodeString err:%s", err)
+		return "", err
+	}
+	var sta state
+	if err := json.Unmarshal(decryptedState, &sta); err != nil {
+		logger.Errorf("Unmarshal err:%s", err)
+		return "", err
+	}
+	codehost, err := GetCodeHost(sta.CodeHostID, logger)
+	if err != nil {
+		return handle(sta.RedirectURL, err)
+	}
+	redirectParsedURL, err := url.Parse(sta.RedirectURL)
+	if err != nil {
+		logger.Errorf("ParseURL:%s err:%s", sta.RedirectURL, err)
+		return "", err
+	}
+	callbackURL := url.URL{
+		Scheme: redirectParsedURL.Scheme,
+		Host:   redirectParsedURL.Host,
+		Path:   callback,
+	}
+	o, err := newOAuth(codehost.Type, callbackURL.String(), codehost.ApplicationId, codehost.ClientSecret, codehost.Address)
+	if err != nil {
+		return handle(sta.RedirectURL, err)
+	}
+	token, err := o.HandleCallback(r)
+	if err != nil {
+		return handle(sta.RedirectURL, err)
+	}
+	codehost.AccessToken = token.AccessToken
+	codehost.RefreshToken = token.RefreshToken
+	if _, err := UpdateCodeHostByToken(codehost, logger); err != nil {
+		logger.Errorf("UpdateCodeHostByToken err:%s", err)
+		return handle(sta.RedirectURL, err)
+	}
+	return handle(sta.RedirectURL, nil)
+}
+
+func newOAuth(provider, callbackURL, clientID, clientSecret, address string) (*oauth.OAuth, error) {
+	switch provider {
+	case systemconfig.GitHubProvider:
+		return oauth.New(callbackURL, clientID, clientSecret, []string{"repo", "user"}, oauth2.Endpoint{
+			AuthURL:  address + "/login/oauth/authorize",
+			TokenURL: address + "/login/oauth/access_token",
+		}), nil
+	case systemconfig.GitLabProvider:
+		return oauth.New(callbackURL, clientID, clientSecret, []string{"api", "read_user"}, oauth2.Endpoint{
+			AuthURL:  address + "/oauth/authorize",
+			TokenURL: address + "/oauth/token",
+		}), nil
+	}
+	return nil, errors.New("illegal provider")
+}
+
+func handle(redirectURL string, err error) (string, error) {
```

Comment:
Use url query to characterize success or failure?

Reply:
yes

Reply:
Is the current interaction logic with the front-end?

Reply:
yes